pkg/dns/dns.go

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright 2024 The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dns
+
+import (
+	"fmt"
+	"net"
+	"slices"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+	"k8s.io/client-go/util/workqueue"
+	"kmesh.net/kmesh/pkg/logger"
+)
+
+var (
+	log = logger.NewLoggerField("ads_controller")
+)
+
+type DNSResolver struct {
+	DnsResolverChan   chan map[string]time.Duration
+	client            *dns.Client
+	resolvConfServers []string
+	cache             map[string]domainCacheEntry
+	// dns refresh priority queue based on exp
+	dnsRefreshQueue workqueue.DelayingInterface
+	sync.RWMutex
+}
+
+// domainCacheEntry stores dns result with expiry time, and also response to trigger dns refresh
+type domainCacheEntry struct {
+	value  []string
+	expiry time.Time
+}
+
+type domainWithRefreshRate struct {
+	name        string
+	refreshRate time.Duration
+}
+
+const MaxConcurrency uint32 = 5
+
+func NewDNSResolver() (*DNSResolver, error) {
+	r := &DNSResolver{
+		DnsResolverChan: make(chan map[string]time.Duration),
+		cache:           map[string]domainCacheEntry{},
+		dnsRefreshQueue: workqueue.NewDelayingQueue(),
+		client: &dns.Client{
+			DialTimeout:  5 * time.Second,
+			ReadTimeout:  5 * time.Second,
+			WriteTimeout: 5 * time.Second,
+		},
+	}
+
+	dnsConfig, err := dns.ClientConfigFromFile("/etc/resolv.conf")
+	if err != nil {
+		return nil, err
+	}
+	if dnsConfig != nil {
+		for _, s := range dnsConfig.Servers {
+			r.resolvConfServers = append(r.resolvConfServers, net.JoinHostPort(s, dnsConfig.Port))
+		}
+	}
+
+	return r, nil
+}
+
+func (r *DNSResolver) StartDNSResolver(stopCh <-chan struct{}) {
+	go r.startResolver()
+	go r.refreshDNS()
+	go func() {
+		<-stopCh
+		r.dnsRefreshQueue.ShutDown()
+		close(r.DnsResolverChan)
+	}()
+}
+
+// startResolver watches the DnsResolver Channel
+func (r *DNSResolver) startResolver() {
+	temp := make(chan struct{}, MaxConcurrency)
+	var wg sync.WaitGroup
+	for domains := range r.DnsResolverChan {
+		temp <- struct{}{}
+		wg.Add(1)
+
+		go func(domains map[string]time.Duration) {
+			defer wg.Done()
+			defer func() { <-temp }()
+			r.resolveDomains(domains)
+		}(domains)
+	}
+	wg.Wait()
+}
+
+// resolveDomains takes a map of hostnames and refresh rate
+func (r *DNSResolver) resolveDomains(domains map[string]time.Duration) {
+	// Stow domain updates, need to remove unwatched domains first
+	r.removeUnwatchedDomain(domains)
```

Comment:
To support removing the UnwatchedDomains,  we must make sure `resolveDomains` should be only be triggered by cds handler. Please add a comment up